OpenTrons/opentrons-api
# fix: send custom labware definitions to the server during LPC setup

## Overview

This PR fixes #9571 by allowing the app to explicitly add labware definitions to the `Run` as part of the LPC setup phase. This ensures that any custom labware loaded during the analysis run will be present for LPC load labware commands, even if those labware definitions were specified at runtime in the Python protocol via `load_labware_by_definition`.

## Changelog

- feat(api-client, react-api-client): add utils to create labware definitions
- feat(robot-server): add POST /runs/:id/labware_definitions endpoint
- feat(app): create labware definitions for custom labware before LPC kicks off

## Review requests

Smoke test procedure for this fix:

1. Upload the smoke test protocol below
2. Ensure LPC works
    - Verify that offsets are actually applied
    - You could do this by jogging to an obviously wrong offset
4. Ensure run works

### Bonus points

1. Modify the smoke test protocol to load the labware from definition onto a module
2. Ensure LPC works
5. Ensure run works

## Risk assessment

Low. If smoke tests prove LPC works, then we're good. None of the changes here affect actual PAPIv2 protocol runs.

## Smoke test protocol

```python
import json
from opentrons import types

metadata = {
    "protocolName": "Test inline custom labware",
    "apiLevel": "2.12",
}

# load_name: fixture_12_trough
# namespace: fixture
# version: 1
TROUGH_DEFINITION = """{"ordering":[["A1"],["A2"],["A3"],["A4"],["A5"],["A6"],["A7"],["A8"],["A9"],["A10"],["A11"],["A12"]],"schemaVersion":2,"version":1,"namespace":"fixture","metadata":{"displayName":"12 Channel Trough","displayVolumeUnits":"mL","displayCategory":"reservoir"},"dimensions":{"xDimension":127.76,"yDimension":85.8,"zDimension":44.45},"parameters":{"format":"trough","isTiprack":false,"isMagneticModuleCompatible":false,"loadName":"fixture_12_trough","quirks":["centerMultichannelOnWells","touchTipDisabled"]},"wells":{"A1":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":13.94,"y":42.9,"z":2.29},"A2":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":23.03,"y":42.9,"z":2.29},"A3":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":32.12,"y":42.9,"z":2.29},"A4":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":41.21,"y":42.9,"z":2.29},"A5":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":50.3,"y":42.9,"z":2.29},"A6":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":59.39,"y":42.9,"z":2.29},"A7":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":68.48,"y":42.9,"z":2.29},"A8":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":77.57,"y":42.9,"z":2.29},"A9":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":86.66,"y":42.9,"z":2.29},"A10":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":95.75,"y":42.9,"z":2.29},"A11":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":104.84,"y":42.9,"z":2.29},"A12":{"shape":"rectangular","depth":42.16,"xDimension":8.33,"yDimension":71.88,"totalLiquidVolume":22000,"x":113.93,"y":42.9,"z":2.29}},"brand":{"brand":"USA Scientific","brandId":["1061-8150"]},"groups":[{"wells":["A1","A2","A3","A4","A5","A6","A7","A8","A9","A10","A11","A12"],"metadata":{"wellBottomShape":"v"}}],"cornerOffsetFromSlot":{"x":0,"y":0,"z":0}}"""  # noqa: E501


def run(ctx):
    tip_rack = ctx.load_labware("opentrons_96_tiprack_300ul", 1)
    pipette = ctx.load_instrument("p300_single", types.Mount.RIGHT, [tip_rack])
    trough = ctx.load_labware_from_definition(
        labware_def=json.loads(TROUGH_DEFINITION),
        location=2,
    )

    pipette.pick_up_tip()
    pipette.aspirate(10, trough.wells()[0].bottom())
    pipette.dispense(10, trough.wells()[1].bottom())
    pipette.return_tip()
```
